src/coreclr/vm/riscv64/stubs.cpp

```diff
@@ -1525,68 +1534,461 @@ static void LoadRegPair(BYTE* p, int reg1, int reg2, UINT32 offset)
+                    // add  t5, a0, t4
+                    *(DWORD*)p = 0x01d50f33; p += 4;
+                    // ld  t5, 0(t5)
+                    *(DWORD*)p = 0x000f3f03; 
+
+                    // move to next indirection offset data
+                    dataOffset = dataOffset - 16 + 4; // subtract 16 as we have moved PC by 16 and add 4 as next data is at 4 bytes from previous data
+                }
+                else
+                {
+                    // ld  t5, #(pLookup->sizeOffset)(a0)
+                    *(DWORD*)p = 0x00053f03 | ((UINT32)pLookup->sizeOffset << 20); p += 4;
+                    dataOffset -= 4; // subtract 4 as we have moved PC by 4
+                }
+                // lui  t4, (slotOffset&0xfffff000)>>12
+                *(DWORD*)p = 0x00000eb7 | ((((UINT32)slotOffset & 0xfffff000) >> 12) << 12); p += 4;
+                // addi  t4, t4, (slotOffset&0xfff)
+                *(DWORD*)p = 0x000e8e93 | ((slotOffset & 0xfff) << 20); p += 4;
+                dataOffset -= 8;
+                // blt  t4, t5, CALL HELPER
+                pBLECall = p;       // Offset filled later
+                *(DWORD*)p = 0x01eec063; p += 4;
+                dataOffset -= 4;
+            }
+
+            if(pLookup->offsets[i] > 2047)
+            {
+                _ASSERTE(dataOffset < 2047);
+                // auipc  t4, 0
+                *(DWORD*)p = 0x00000297;
+                p += 4;
+                // lw  t4, dataOffset(t4)
+                *(DWORD*)p = 0x000eae83 | ((dataOffset & 0xfff) << 20);
+                p += 4;
+                // add a0, a0, t4
+                *(DWORD*)p = 0x01d50533; 
+                // lw  a0, 0(a0)
+                *(DWORD*)p = 0x00052503; 
+                p += 4;
+                // move to next indirection offset data
+                dataOffset = dataOffset - 16 + 4; // subtract 16 as we have moved PC by 16 and add 4 as next data is at 4 bytes from previous data
+            }
+            else
+            {
+                // offset must be 8 byte aligned
+                _ASSERTE((pLookup->offsets[i] & 0x7) == 0);
+                // ld  a0, #(pLookup->offsets[i])(a0)
+                *(DWORD*)p = 0x00053503 | ((UINT32)pLookup->offsets[i] << 20);
+                p += 4;
+                dataOffset -= 4; // subtract 4 as we have moved PC by 4
+            }
+        }
+
+        // No null test required
+        if (!pLookup->testForNull)
+        {
+            _ASSERTE(pLookup->sizeOffset == CORINFO_NO_SIZE_CHECK);
+            // ret
+            *(DWORD*)p = 0x00008067;
+            p += 4;
+        }
+        else
+        {
+            // beq  a0, x0, CALL HELPER:
+            *(DWORD*)p = 0x00050063;
+            p += 4;
+
+            // ret
+            *(DWORD*)p = 0x00008067;
+            p += 4;
+
+            // CALL HELPER:
+            if(pBLECall != NULL)
+                *(DWORD*)pBLECall |= ((UINT32)(p - pBLECall) << 8);
```

Comment:
I think this sets offset to pBLECall (blt t4, t5, CALL HELPER). Please check the below how offset is splitted into instruction.
```
// B-immediate encodes a signed offset in multiples of 2 bytes
//       12      | 11                               1 | 0
// inst[31]/sign | inst[7] | inst[30:25] | inst[11:8] | 0
```
\+ Please add an assertion for offset.

Reply:
I see why you shift only once. It is because max value of `p - pBLECall` under 5 bits. However, I think shift value should be '7' not '8'. And please add assertions (0th is 0 and max value is under 5 bits) for value of `p - pBLECall`.

Reply:
Done